src/linearalgebra.jl

```diff
@@ -247,54 +247,58 @@ function LinearAlgebra.eigen(T::Hermitian{ElT,<:DenseTensor{ElT,2,IndsT}};
 end
 
 """
-    random_orthog(n::Int,m::Int)
+    random_unitary(n::Int,m::Int)::Matrix{ComplexF64}
+    random_unitary(::Type{ElT},n::Int,m::Int)::Matrix{ElT}
 
-Return a random matrix O of dimensions (n,m)
-such that if n >= m, transpose(O)*O is the 
-identity, or if m > n O*transpose(O) is the
-identity.
-"""
-function random_orthog(n::Int,m::Int)::Matrix
-  #TODO: optimize by using recursive Householder algorithm?
-  if n < m
-    return transpose(random_orthog(m,n))
-  end
-  F = qr(randn(n,m))
-  Q = Matrix(F.Q)
-  for c=1:size(Q,2)
-    if real(F.R[c,c]) < 0.0
-      Q[:,c] *= -1
-    end
-  end
-  return Q
-end
-
-"""
-    random_unitary(n::Int,m::Int)::Matrix
-
-Return a random complex matrix U of dimensions (n,m)
-such that if n >= m, U'*U is the
-identity, or if m > n U*U' is the
-identity.
+Return a random matrix U of dimensions (n,m)
+such that if n >= m, U'*U is the identity, or if 
+m > n U*U' is the identity. Optionally can pass a numeric
+type as the first argument to obtain a matrix of that type.
 
 Sampling is based on https://arxiv.org/abs/math-ph/0609050
 such that in the case `n==m`, the unitary matrix will be sampled
 according to the Haar measure.
 """
-function random_unitary(n::Int,m::Int)
+function random_unitary(::Type{ElT}, n::Int,m::Int) where {ElT <: Number}
   if n < m
-    return Matrix(random_unitary(m,n)')
+    return Matrix(random_unitary(ElT,m,n)')
   end
-  F = qr(randn(ComplexF64,n,m))
+  F = qr(randn(ElT,n,m))
   Q = Matrix(F.Q)
   A = diag(Matrix(F.R))
-  A ./= abs.(A)
+  pn(x) = (real(x) < 0.0 ? -1.0 : 1.0)
+  A .= pn.(A)
```

Comment:
Doesn't the old version still work in the complex case?

Reply:
I made this new version because I got a bit nervous about seeing a division happening there. What if one of the values in A happened to be very small or even zero?

Reply:
So to clarify one other thing, this new version wasn't about the complex case per se. I just implemented the new version wrong at first, leaving out the `real(x)` and just putting `x`, so there was another commit fixing that.

Reply:
I see. It looks like we could use the Julia function `sign` for this, so replace `A ./= abs.(A)` with `A .= sign.(A)`:
```julia
help?> sign
search: sign signed Signed signbit significand Unsigned unsigned flipsign copysign isassigned

  sign(x)


  Return zero if x==0 and x/|x| otherwise (i.e., ±1 for real x).
```
Though maybe for this case we want the behavior `sign(0) == 1`? i.e.:
```julia
_sign(x) = iszero(x) ? one(x) : sign(x)
```
then use it as `A .= _sign.(A)`.